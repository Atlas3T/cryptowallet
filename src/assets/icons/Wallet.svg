<?xml version="1.0" encoding="utf-8"?>
<!-- Generator: Adobe Illustrator 22.1.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 viewBox="0 0 51 51" style="enable-background:new 0 0 51 51;" xml:space="preserve">
<style type="text/css">
	.st0{fill:none;stroke:#000000;stroke-width:5;stroke-miterlimit:10;}
	.st1{display:none;}
	.st2{display:inline;fill:none;stroke:#000000;stroke-linecap:round;stroke-linejoin:round;stroke-miterlimit:10;}
	.st3{display:inline;}
</style>
<g id="Wallet">
	<path class="st0" d="M46.1,31.5v5.3c0,2.4-1.9,4.3-4.3,4.3H9.5c-2.4,0-4.3-1.9-4.3-4.3V15.5c0-2.4,1.9-4.3,4.3-4.3h32.3
		c2.4,0,4.3,1.9,4.3,4.3v5.2"/>
	<circle cx="46.1" cy="26.1" r="3.1"/>
</g>
<g id="Payments" class="st1">
	<line class="st2" x1="8" y1="19.8" x2="41.8" y2="19.8"/>
	<polyline class="st2" points="35.8,13.3 42.3,19.8 35.8,26.3 	"/>
	<line class="st2" x1="42.3" y1="32.8" x2="8.5" y2="32.8"/>
	<polyline class="st2" points="14.5,39.3 8,32.8 14.5,26.3 	"/>
</g>
<g id="Exchange" class="st1">
	<path class="st2" d="M15.3,21.7c0-5.6,4.6-10.2,10.2-10.2s10.2,4.6,10.2,10.2"/>
	<polyline class="st2" points="42.2,16.3 35.7,22.8 29.2,16.3 	"/>
	<path class="st2" d="M35.7,30.2c0,5.6-4.6,10.2-10.2,10.2s-10.2-4.6-10.2-10.2"/>
	<polyline class="st2" points="8.8,35.7 15.3,29.2 21.9,35.7 	"/>
</g>
<g id="More" class="st1">
	<circle class="st3" cx="26" cy="15.8" r="7.8"/>
	<circle class="st3" cx="12.3" cy="28.6" r="5"/>
	<circle class="st3" cx="35.3" cy="33.6" r="3.5"/>
</g>
</svg>
