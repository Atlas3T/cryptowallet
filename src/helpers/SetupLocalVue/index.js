import Vuex from 'vuex';
import { createLocalVue } from '@vue/test-utils';
import VueRouter from 'vue-router';
import VueI18n from 'vue-i18n';
import messages from '@/i18n/';
import axios from 'axios';
import { sync } from 'vuex-router-sync';
import errorHandlerPlugin from '@/boot/ErrorHandler';
import toasterPlugin from '@/boot/Toaster';
import permissionsPlugin from '@/boot/Permissions';
import appInvitationPlugin from '@/boot/AppInvitation';
import smsPlugin from '@/boot/Sms';
import emailPlugin from '@/boot/Email';
import contactsImportPlugin from '@/boot/ContactsImport';
import * as All from 'quasar'

const { Quasar, date } = All;

const components = Object.keys(All).reduce((object, key) => {
  const val = All[key]
  if (val && val.component && val.component.name != null) {
    object[key] = val
  }
  return object
}, {})


const localVue = createLocalVue();

localVue.use(Quasar, { components });
localVue.use(VueI18n);
localVue.use(Vuex);
// localVue.use(VueRouter);
localVue.prototype.$axios = axios;
errorHandlerPlugin({ Vue: localVue });
toasterPlugin({ Vue: localVue });
permissionsPlugin({ Vue: localVue });
appInvitationPlugin({ Vue: localVue });
smsPlugin({ Vue: localVue });
emailPlugin({ Vue: localVue });
contactsImportPlugin({ Vue: localVue });

/**
 * creates a new router and syncs it with the store using vuex-router-sync
 * @param store
 * @returns router
 */
const createRouter = (store = null) => {
  const router = new VueRouter({
    routes: [
      {
        path: '/contact/:id',
      },
    ],
  });

  if (store) { sync(store, router); }
  return router;
};

const i18n = new VueI18n({ messages, locale: 'en-gb' });

export {
  localVue,
  createRouter,
  i18n,
};
