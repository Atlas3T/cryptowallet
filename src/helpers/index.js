import AmountFormatter from './AmountFormatter';
import dateTranslater from './DateTranslater';
import Contacts from './Contacts';

export {
  AmountFormatter,
  dateTranslater,
  Contacts,
};
