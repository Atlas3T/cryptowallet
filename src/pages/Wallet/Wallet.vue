<template>
  <div class="layout">
    <div class="list">
      <WalletsList
        :coins="coins"
        :currency="currency"
      />
    </div>
    <div class="button">
      <AddWallet/>
    </div>
  </div>
</template>

<script>
import WalletsList from '@/components/Wallet/Wallet/WalletsList.vue';
import AddWallet from '@/components/Wallet/Wallet/AddWallet.vue';

export default {
  name: 'Balance',
  components: {
    WalletsList,
    AddWallet,
  },
  /**
   * @TODO James Konrad
   * data below is mocked
   *
   * currency String is a prefered user currency, should be in the Vuex store
   * you should probably select your prefered currency during the initial setup
   * alongside the language
   *
   * coins Object also shoudn't be here
   */
  data() {
    return {
      currency: 'GPB',
      coins: {
        BTC: {
          key: 'BTC',
          name: 'Bitcoin',
          balance: 25.532344,
          uid: 1,
        },
        ATL: {
          key: 'ATL',
          name: 'Atlas',
          balance: 21.532344,
          uid: 2,
        },
        ETH: {
          key: 'ETH',
          name: 'Ethereum',
          balance: 55.532344,
          uid: 3,
        },
      },
    };
  },
};

</script>

<style scoped>
.layout {
  display: flex;
  flex-direction: column;
}

.list {
  flex: 1;
  position: relative;
}
</style>
