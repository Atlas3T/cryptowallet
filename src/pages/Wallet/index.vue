<template>
  <div class="layout">
    <div class="list">
      <wallet-list
        :wallets="wallets"
        :currency="selectedCurrency"
      />
    </div>
    <div class="button">
      <add-wallet/>
    </div>
  </div>
</template>

<script>
import WalletList from '@/components/Wallet/WalletsList';
import AddWallet from '@/components/Wallet/AddWallet';

export default {
  name: 'Balance',

  components: {
    WalletList,
    AddWallet,
  },

  /**
   * @TODO James Konrad
   * data below is mocked
   *
   * currency String is a prefered user currency, should be in the Vuex store
   * you should probably select your prefered currency during the initial setup
   * alongside the language
   *
   * coins Object also shoudn't be here
   */
  data() {
    return {
      wallets: {
        BTC: {
          key: 'BTC',
          name: 'Bitcoin',
          balance: 25.532344,
          uid: 1,
        },
        ATL: {
          key: 'ATL',
          name: 'Atlas',
          balance: 21.532344,
          uid: 2,
        },
        ETH: {
          key: 'ETH',
          name: 'Ethereum',
          balance: 55.532344,
          uid: 3,
        },
      },
    };
  },

  computed: {
    selectedCurrency() {
      return this.$store.state.settings.selectedCurrency;
    },
  },
};

</script>

<style scoped>
.layout {
  display: flex;
  flex-direction: column;
}

.list {
  flex: 1;
  position: relative;
}
</style>
