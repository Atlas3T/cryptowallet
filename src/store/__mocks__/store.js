import Vue from 'vue';
import Vuex from 'vuex';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from '@atlascity/vuex-orm-plugin-lokijs';

import wallet from '@/store/wallet';
import contacts from '@/store/contacts/__mocks__/contacts.js';
import search from '@/store/search/__mocks__/search.js';
import settings from '@/store/settings/__mocks__/settings.js';
import setup from '@/store/setup/__mocks__/setup.js';
import qrcode from '@/store/qrcode/__mocks__/qrcode.js';
import modals from '@/store/modals/__mocks__/modals.js';

// entities
import Account from '@/store/wallet/entities/account';
import Address from '@/store/wallet/entities/address';
import Tx from '@/store/wallet/entities/tx';
import Utxo from '@/store/wallet/entities/utxo';
import Wallet from '@/store/wallet/entities/wallet';
import Coin from '@/store/wallet/entities/coin';
import KeyPair from '@/store/wallet/entities/keyPair';
import Prices from '@/store/prices';
import LatestPrice from '@/store/latestPrice';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
  Wallet,
];

const options = {
  env: 'browser',
};

VuexORM.use(VuexORMLoki, { database, options });

export const actions = {};
export const getters = {};
export const mutations = {};
export const state = {};

export function createMocks(custom = {
  getters: {}, mutations: {}, actions: {}, state: {},
}) {
  /**
   * mock store modules.
   */
  const modules = {
    wallet,
    search,
    settings,
    contacts,
    setup,
    qrcode,
    modals,
  };

  let mockGettersCombined = {};
  let mockMutationsCombined = {};
  let mockActionsCombined = {};
  let mockStateCombined = {};

  Object.entries(modules).forEach(([name, module]) => {
    mockGettersCombined = Object.assign({}, mockGettersCombined, module.getters);
    mockMutationsCombined = Object.assign({}, mockMutationsCombined, module.mutations);
    mockActionsCombined = Object.assign({}, mockActionsCombined, module.actions);
    mockStateCombined = Object.assign({}, mockStateCombined, module.state);

    if (custom.state && custom.state[name]) {
      module.state = Object.assign(module.state, custom.state[name]);
    }

    modules[name] = {
      namespaced: true,
      getters: Object.assign({}, getters, module.getters),
      mutations: Object.assign({}, mutations, module.mutations),
      actions: Object.assign({}, actions, module.actions),
      state: Object.assign({}, state, module.state),
    };
  });

  mockGettersCombined = Object.assign({}, mockGettersCombined, custom.getters);
  mockMutationsCombined = Object.assign({}, mockMutationsCombined, custom.mutations);
  mockActionsCombined = Object.assign({}, mockActionsCombined, custom.actions);
  mockStateCombined = Object.assign({}, mockStateCombined, custom.state);
  return {
    getters: mockGettersCombined,
    mutations: mockMutationsCombined,
    actions: mockActionsCombined,
    state: mockStateCombined,
    store: new Vuex.Store({
      getters: mockGettersCombined,
      mutations: mockMutationsCombined,
      actions: mockActionsCombined,
      state: mockStateCombined,
      modules,
      plugins: [VuexORM.install(database)],
    }),
  };
}

export const { store } = createMocks();
