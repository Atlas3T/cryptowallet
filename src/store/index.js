/*eslint-disable*/

import Vue from 'vue';
import Vuex from 'vuex';
import VuexPersistence from 'vuex-persist';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from 'vuex-orm-lokijs';

// import entities.
import Account from './wallet/entities/account';
import Address from './wallet/entities/address';
import Tx from './wallet/entities/tx';
import Utxo from './wallet/entities/utxo';
import Wallet from './wallet/entities/wallet';
import Coin from './wallet/entities/coin';
import KeyPair from './wallet/entities/keyPair';
import Prices from './prices';
import LatestPrice from './latestPrice';

// import modules.
import payments from './payments';
import search from './search';
import settings from './settings';
import setup from './setup';
import contacts from './contacts';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
];

window.Tx = Tx;
window.Utxo = Utxo;
window.Address = Address;
window.Wallet = Wallet;
window.LatestPrice = LatestPrice;
window.Prices = Prices;
window.Account = Account;
window.Coin = Coin;

// Setup persistant storage.
const vuexLocal = new VuexPersistence({
  supportCircular: true,
  storage: window.localStorage,
  modules: [settings],
});


const options = {
  env: 'browser',
};

function hydrationCompletedCallback() {
  setTimeout(() => {
    store.dispatch('settings/setLoading', false);
  }, 1000);
}

VuexORM.use(VuexORMLoki, { database, options, hydrationCompletedCallback });


/**
 * Create CryptoWallet Vuex store obj.
 * @type {Store<any>}
 */
const store = new Vuex.Store({
  modules: {
    payments,
    search,
    settings,
    setup,
    contacts,
  },
  plugins: [VuexORM.install(database)],
});

Vue.prototype.activeWallets = {};

// @todo remove line below (used for development purposes only)
window.store = store;

export default store;
