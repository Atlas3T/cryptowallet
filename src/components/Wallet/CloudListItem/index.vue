<template>
  <div
    class="wallet-cloud"
    @click="goToWallet"
  >
    <div class="cloud-top-row">
      <CoinHeader :wallet="wallet" />
    </div>
    <div
      v-if="chartData"
      class="trend-wrapper"
    >
      <trend
        :data="chartData"
        :gradient="['#fabc57']"
        :stroke-width="5"
        auto-draw
        smooth
      />
    </div>
    <div class="wallet-buttons">
      <q-btn-group>
        <q-btn
          :disabled="cantSend"
          icon="send"
          size="md"
          color="primary"
          label="Send"
          class="wallet-group-btn"
          flat
          @click.stop="send"
        />
        <q-btn
          icon="call_received"
          size="md"
          color="primary"
          label="Receive"
          class="wallet-group-btn"
          flat
          @click.stop="receive"
        />
        <q-btn
          v-if="showChart"
          size="lg"
          color="primary"
          icon="timeline"
          class="wallet-group-btn just-icon"
          flat
          @click.stop="prices"
        />
      </q-btn-group>
    </div>
  </div>
</template>

<script>
import { mapState } from 'vuex';
import CoinHeader from '@/components/Wallet/CoinHeader';
import Prices from '@/store/prices';
import { getBalance } from '@/helpers';

export default {
  name: 'CloudListItem',
  components: {
    CoinHeader,
  },
  props: {
    wallet: {
      type: Object,
      required: true,
    },
    currency: {
      type: Object,
      required: false,
    },
  },
  data() {
    return {
      chartData: null,
    };
  },
  computed: {
    ...mapState({
      authenticatedAccount: (state) => {
        return state.settings.authenticatedAccount;
      },
    }),
    selectedCurrency() {
      return this.$store.state.settings.selectedCurrency;
    },
    showChart() {
      const chartData = this.$store.getters['entities/prices/find'](`${this.wallet.symbol}_${this.selectedCurrency.code}_day`);
      if (chartData) {
        return true;
      }
      return false;
    },
    cantSend() {
      return getBalance(this.wallet, this.authenticatedAccount).available === 0;
    },
  },
  async mounted() {
    try {
      const hour = 3600000;
      const currentTime = new Date().getTime();
      const price = Prices.find([`${this.wallet.symbol}_${this.selectedCurrency.code}_day`]);

      if (price) {
        if ((currentTime - price.updated) < hour) {
          this.chartData = price.data.map((item) => { return item.y; });
        } else {
          let dataset;
          const result = await this.backEndService.getHistoricalData(this.wallet.symbol, this.selectedCurrency.code, 'day');
          if (result) {
            dataset = result.data;
          }
          if (dataset) {
            this.backEndService.storeChartData(this.wallet.symbol, 'day', dataset);
            this.chartData = dataset.map((item) => { return item.y; });
          } else {
            this.chartData = price.data.map((item) => { return item.y; });
          }
        }
      }
    } catch (err) {
      this.errorHandler(err);
    }
  },
  methods: {
    send() {
      this.$router.push({ path: `/wallet/send/${this.wallet.id}` });
    },
    receive() {
      this.$router.push({ path: `/wallet/receive/${this.wallet.id}` });
    },
    prices() {
      this.$router.push({ path: `/wallet/prices/${this.wallet.id}` });
    },
    goToWallet() {
      this.$router.push({ path: `/wallet/single/${this.wallet.id}` });
    },
  },
};
</script>

<style>
.wallet-cloud {
  width: 100%;
  border-radius: 0.4em;
  padding-bottom: 0.2em;
  background: white;
  color: #1e3c57;
  /* border: 1px solid rgba(0, 0, 0, 0.03); */
  margin-bottom: 0.5em;
  box-shadow: inset -2px -2px 0px 2px rgba(29, 59, 86, 0.35);
}

.cloud-top-row {
  padding: 1em 1em 0em 1em;
}

.trend-wrapper {
  opacity: 0.7;
  margin: .5em 0;
  padding: 0 1em;
}

.wallet-buttons {
  border-top: 1px solid #f2f3f5;
  margin-right: 0.25rem;
  margin-top: 1rem;
}

.wallet-buttons .q-btn-group {
  width: 100%;
  display: flex;
  align-items: stretch;
}

.wallet-buttons .q-btn {
  display: block;
  flex: 1;
  white-space: nowrap;
}

.wallet-buttons .q-btn__content {
  font-size: 1.1em;
  color: #1e3c57;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
}

.wallet-group-btn {
  background: transparent;
  padding: 0 .5rem;
}

.wallet-group-btn i {
  margin-right: 0.5em;
  position: relative;
  top: -0.05rem;
}

.wallet-group-btn.just-icon {
  min-height: 2.5rem!important;
}

.wallet-group-btn.just-icon i {
  margin-right: 0;
  font-size: 2rem;
  text-align: right;
}

.wallet-group-btn .q-btn__content {
  font-size: 0.8rem;
}
</style>
