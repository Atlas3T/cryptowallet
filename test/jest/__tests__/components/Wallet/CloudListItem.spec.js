/* eslint-disable no-magic-numbers */
import CloudListItem from '@/components/Wallet/CloudListItem';
import { mount } from '@vue/test-utils';
import { localVue, i18n, createRouter } from '@/helpers/SetupLocalVue';
import { createMocks as createStoreMocks } from '@/store/__mocks__/store.js';
import Wallet from '@/store/wallet/entities/wallet';
import Prices from '@/store/prices';

describe('CloudListItem.vue', () => {
  let storeMocks;
  let wrapper;
  let router;

  const chartData = JSON.parse('{"$id":"CAT_GBP_day","coin":"CAT","currency":"GBP","period":"day","updated":15580,"data":[{"t":15560100,"y":135.02},{"t":15560136,"y":134.7},{"t":15560172,"y":134.33},{"t":15560208,"y":134.57},{"t":15560244,"y":135.14},{"t":15560280,"y":135.43},{"t":15560316,"y":134.65},{"t":15560352,"y":134.81},{"t":15560388,"y":134.8},{"t":15560424,"y":134.79},{"t":15560460,"y":134.12},{"t":15560496,"y":133.71},{"t":15560532,"y":134.17},{"t":15560568,"y":132.86},{"t":15560604,"y":132.23},{"t":15560640,"y":132.27},{"t":15560676,"y":131.07},{"t":15560712,"y":131.49},{"t":15560748,"y":131.61},{"t":15560784,"y":131.92},{"t":15560820,"y":131.59},{"t":15560856,"y":131.29},{"t":15560892,"y":128.03},{"t":15560928,"y":128.1},{"t":15560964,"y":128.02}]}');

  const errorHandler = jest.fn();

  const backEndService = {
    getHistoricalData: jest.fn(),
    storeChartData: jest.fn(),
  };

  const propsData = {
    wallet: {
      id: 1,
      name: 'Catalyst',
      displayName: 'Catalyst',
      parentName: 'Ethereum',
      symbol: 'CAT',
      network: 'ETHEREUM_ROPSTEN',
      sdk: 'ERC20',
      confirmedBalance: 20,
    },
  };

  function wrapperInit(options) {
    return mount(CloudListItem, options);
  }

  function storeInit(custom) {
    storeMocks = createStoreMocks(custom);
    Prices.$insert({ data: chartData });
    Wallet.$insert({
      data: [{
        id: 1,
        account_id: 1,
        name: 'Catalyst',
        displayName: 'Catalyst',
        parentName: 'Ethereum',
        symbol: 'CAT',
        network: 'ETHEREUM_ROPSTEN',
        sdk: 'ERC20',
        confirmedBalance: 20,
        enabled: true,
        imported: true,
      },
      {
        id: 2,
        account_id: 1,
        name: 'Ethereum',
        displayName: 'Ethereum',
        symbol: 'ETH',
        network: 'ETHEREUM_ROPSTEN',
        sdk: 'ERC20',
        confirmedBalance: 20,
        enabled: true,
        imported: true,
      }],
    });
    router = createRouter(storeMocks.store);
    router.push({ path: '/wallet' });
    wrapper = wrapperInit({
      i18n,
      router,
      localVue,
      store: storeMocks.store,
      propsData,
      mocks: { errorHandler, backEndService },
      stubs: { CoinHeader: true, trend: true },
    });
  }

  it('renders and matches snapshot', () => {
    storeInit();
    expect(wrapper.element).toMatchSnapshot();
  });

  it('displays the price chart button if data is available', (done) => {
    backEndService.getHistoricalData.mockImplementation(() => {
      return {
        code: 'CAT', timestamp: 1556100003, currency: 'GBP', period: 'day', data: [{ t: 15560136, y: 134.7 }, { t: 15560172, y: 134.33 }, { t: 15560208, y: 134.57 }, { t: 15560244, y: 135.14 }, { t: 15560280, y: 135.43 }, { t: 15560316, y: 134.65 }, { t: 15560352, y: 134.81 }, { t: 15560388, y: 134.8 }, { t: 15560424, y: 134.79 }, { t: 1556046, y: 134.12 }, { t: 15560496, y: 133.71 }, { t: 15560532, y: 134.17 }, { t: 15560568, y: 132.86 }, { t: 15560604, y: 132.23 }, { t: 15560640, y: 132.27 }, { t: 15560676, y: 131.07 }, { t: 15560712, y: 131.49 }, { t: 15560748, y: 131.61 }, { t: 15560784, y: 131.92 }, { t: 15560820, y: 131.59 }, { t: 15560856, y: 131.29 }, { t: 15560892, y: 128.03 }, { t: 15560928, y: 128.1 }, { t: 15560964, y: 127.61 }, { t: 15561000, y: 127.13 }],
      };
    });
    storeInit();
    Prices.$insert({ data: chartData });
    setTimeout(() => {
      expect(wrapper.findAll('button').at(2).text()).toBe('timeline');
      done();
    }, 0);
  });

  it('calls the api for chart data and updates the database', (done) => {
    storeInit();
    setTimeout(() => {
      expect(backEndService.storeChartData).toHaveBeenCalled();
      done();
    }, 501);
  });

  it('hides the chart if no price data exists', (done) => {
    backEndService.getHistoricalData.mockImplementationOnce(() => {});
    storeInit();
    Prices.$delete('CAT_GBP_day');
    setTimeout(() => {
      expect(wrapper.vm.showChart).toEqual(false);
      done();
    }, 0);
  });

  it('does not display a price chart if no data exists and it cannot be fetched', (done) => {
    storeInit();
    setTimeout(() => {
      expect(Prices.all().length).toEqual(1);
      done();
    }, 0);
  });

  it('handles any errors', (done) => {
    backEndService.getHistoricalData.mockImplementationOnce(() => { throw new Error('Test Error'); });
    storeInit();
    setTimeout(() => {
      expect(errorHandler).toHaveBeenCalledWith(Error('Test Error'));
      done();
    }, 0);
  });

  it('does not fetch new data if data is less than an hour old', (done) => {
    chartData.updated = Date.now();
    storeInit();
    setTimeout(() => {
      expect(wrapper.vm.chartData[0]).toEqual(135.02);
      done();
    }, 0);
  });


  it('the send button routes to /wallet/send', (done) => {
    storeInit();
    wrapper.findAll('button').at(0).trigger('click');
    setTimeout(() => {
      expect(router.history.current.path).toBe('/wallet/send/1');
      done();
    }, 0);
  });

  it('the receive button routes to /wallet/receive', (done) => {
    storeInit();
    wrapper.findAll('button').at(1).trigger('click');
    setTimeout(() => {
      expect(router.history.current.path).toBe('/wallet/receive/1');
      done();
    }, 0);
  });

  it('the chart button routes to /wallet/prices', (done) => {
    storeInit();
    Prices.$insert({ data: chartData });
    wrapper.findAll('button').at(2).trigger('click');
    setTimeout(() => {
      expect(router.history.current.path).toBe('/wallet/prices/1');
      done();
    }, 0);
  });

  it('the cloud item routes to /wallet/single', (done) => {
    storeInit();
    wrapper.find('.wallet-cloud').trigger('click');
    setTimeout(() => {
      expect(router.history.current.path).toBe('/wallet/single/1');
      done();
    }, 0);
  });
});
