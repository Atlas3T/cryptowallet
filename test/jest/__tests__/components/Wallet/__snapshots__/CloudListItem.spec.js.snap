// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletItem.vue renders and matches snapshot 1`] = `
<div
  class="wallet-cloud"
>
  <div
    class="cloud-top-row"
  >
    <coinheader-stub
      wallet="[object Object]"
    />
  </div>
   
  <!---->
   
  <div
    class="wallet-buttons"
  >
    <div
      class="q-btn-group row no-wrap inline"
    >
      <button
        class="q-btn inline q-btn-item non-selectable wallet-group-btn q-btn--flat q-btn--rectangle text-primary q-focusable q-hoverable"
        style="font-size: 14px;"
        tabindex="0"
        type="button"
      >
        <div
          class="q-focus-helper"
          tabindex="-1"
        />
        <div
          class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
        >
          <i
            aria-hidden="true"
            class="material-icons q-icon on-left"
          >
            send
          </i>
          <div>
            Send
          </div>
        </div>
      </button>
       
      <button
        class="q-btn inline q-btn-item non-selectable wallet-group-btn q-btn--flat q-btn--rectangle text-primary q-focusable q-hoverable"
        style="font-size: 14px;"
        tabindex="0"
        type="button"
      >
        <div
          class="q-focus-helper"
          tabindex="-1"
        />
        <div
          class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
        >
          <i
            aria-hidden="true"
            class="material-icons q-icon on-left"
          >
            call_received
          </i>
          <div>
            Receive
          </div>
        </div>
      </button>
       
      <!---->
    </div>
  </div>
</div>
`;
